File: Python/maximum-requests-without-violating-the-limit.py
# Time:  O(nlogn)
# Space: O(n)

import collections


# hash table, sort, deque, sliding window
class Solution(object):
    def maxRequests(self, requests, k, window):
        """
        :type requests: List[List[int]]
        :type k: int
        :type window: int
        :rtype: int
        """
        lookup = collections.defaultdict(list)
        for u, t in requests:
            lookup[u].append(t)
        result = len(requests)
        for l in lookup.itervalues():
            l.sort()
            dq = collections.deque()
            for x in l:
                dq.append(x)
                if len(dq) <= k:
                    continue
                if dq[-1]-dq[0] > window:
                    dq.popleft()
                    continue
                dq.pop()
                result -= 1                    
        return result

File: C++/maximum-requests-without-violating-the-limit.cpp
// Time:  O(nlogn)
// Space: O(n)

// hash table, sort, deque, sliding window
class Solution {
public:
    int maxRequests(vector<vector<int>>& requests, int k, int window) {
        unordered_map<int, vector<int>> lookup;
        for (const auto& r : requests) {
            lookup[r[0]].emplace_back(r[1]);
        }
        int result = size(requests);
        for (auto& [_, l] : lookup) {
            ranges::sort(l);
            deque<int> dq;
            for (const auto& x : l) {
                dq.emplace_back(x);
                if (size(dq) <= k) {
                    continue;
                }
                if (dq.back() - dq.front() > window) {
                    dq.pop_front();
                    continue;
                }
                dq.pop_back();
                --result;
            }
        }
        return result;
    }
};
